sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java

```diff
@@ -334,6 +334,10 @@ public void setAttribute(String key, AttributeValue value) {
     Preconditions.checkNotNull(key, "key");
     if (value == null || (value.getType().equals(STRING) && value.getStringValue() == null)) {
       synchronized (lock) {
+        if (hasEnded) {
```

Comment:
why don't we just do this check once up front, rather than repeating the code here and below?

Reply:
I didn't want to introduce another `synchronized` block. But I guess you're right, it is better to do only once upfront!

Reply:
I'd rather have one slightly larger synchronized block, rather than force jumping in and out of the synchronization twice in rapid succession. As long as you're doing very cheap operations, just putting all the logic in one synchronization context is going to be better.

Reply:
But this one does not jump between synchronized.

Reply:
I agree. I was just thinking forward to moving the ended check up-front, rather then duplicating that check.

Reply:
Splitting the lock is incorrect. You need to make the check in the same lock sync to ensure that the condition does not change between the two synchronized sections

Reply:
To clarify @jkwatson your proposal is correct but your concern should not be a preference should be a requirement because otherwise the code is incorrect :)

Reply:
You are right @bogdandrutu, I was thinking only to the case where the span already ended.

Reply:
@bogdandrutu also true! I hadn't actually fully thought through the implementation, but you are absolutely correct.

Reply:
Personally I am happy with the current code, but not opposed to change it based on @jkwatson suggestion. Up to you @thisthat :)

Reply:
I think having everything in a single synchronized block makes it easier to fully understand the method. 7cab1ee addressed this :wink: